### PR TITLE
DAOS-9134 control: prevent simultaneous spdk operations on single device

### PR DESCRIPTION
**Pull request by tanabarr**
During engine start-up when VMD is enabled, VFIO resource busy errors
can occur. This can happen in a race scenario where a remote dmg storage
scan is triggered at the same time as either a format operation or
engine start.

Solution is to:
* protect storage provider operations with mutex (RWMutex->Mutex)
* avoid using SPDK bindings for scan as engine processes are started

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

**Comment by NiuYawei** [comment]
Is this a VMD specific issue? Can normal device allow concurrent access from both daos server and io engine? I was thinking daos server and io engine always accesses NVMe SSD exclusively.

File: src/control/server/storage/provider.go [hunk 1]
```diff
@@ -288,11 +288,11 @@ func (p *Provider) FormatBdevTiers() (results []BdevTierFormatResult) {
 			continue
 		}
 
-		p.RLock()
+		p.Lock()
 		req.BdevCache = &p.bdevCache
```
Comment (kjacque) [inline src/control/server/storage/provider.go:292]: I found myself wondering if perhaps the spdk operations should be locking at a lower level, for example in the bdev provider itself. At this level it's hard to tell what exactly the conflict was. Maybe I'm wrong but on re-visiting this code, this lock appears to be originally intended to protect the bdev cache.

I'm not totally against expanding its purpose, but it seems like "spdk operations can conflict" is bdev provider specific knowledge.

Comment (tanabarr) [reply src/control/server/storage/provider.go:292]: For the same reason as protecting the bdev cache, it needs to be at this level (storage provider) in order to persist across daos_admin process reincarnations. Locking in the bdev provider backend wouldn't protect against multiple invocations from the storage provider level. Make sense?

Comment (kjacque) [reply src/control/server/storage/provider.go:292]: Ah, understood.

**Review by kjacque — DISMISSED** [review]

**Review by mjmac — COMMENTED** [review]
Would prefer to simplify this PR to just address the core issue which is that the storage provider needs to prevent concurrent SPDK operations.

I am also not convinced that we should be using locks this way. Locking is easy to add but can be difficult to get correct, particularly when there are multiple entry points taking locks at different times and places. For example, `PrepareBdevs()` was not modified in this PR and I believe it is now unprotected, as the intended use of the mutex has changed.

I think you can achieve your goal more simply and reliably by overriding the `SendReq()` method on `BdevAdminForwarder`, e.g.
```go
func (f *BdevAdminForwarder) SendReq(method string, fwdReq interface{}, fwdRes interface{}) error {
	f.reqMutex.Lock()
	defer f.reqMutex.Unlock()

	return f.Forwarder.SendReq(method, fwdReq, fwdRes)
}
```

This way, you've only got one place to maintain the locking, and there's no way for someone to mistakenly add new code that bypasses it.

**Comment by tanabarr** [comment]
> Is this a VMD specific issue? Can normal device allow concurrent access from both daos server and io engine? I was thinking daos server and io engine always accesses NVMe SSD exclusively.

yes this is a VMD specific issue, it doesn't occur during non-VMD operation as far as I know

**Comment by tanabarr** [comment]
@mjmac as requested I have split out the concurrency specific fixes into https://github.com/daos-stack/daos/pull/7519 and will leave the peripheral changes in this PR.

**Comment by tanabarr** [comment]
separated out into 2:
- https://github.com/daos-stack/daos/pull/7530)
- https://github.com/daos-stack/daos/pull/7519

**Pull request closed without merging**